src/rosconsole/impl/rosconsole_log4cxx.cpp

```diff
@@ -376,14 +375,14 @@ void shutdown()
   {
     const log4cxx::LoggerPtr& logger = log4cxx::Logger::getLogger(ROSCONSOLE_ROOT_LOGGER_NAME);
     logger->removeAppender(g_log4cxx_appender);
-    g_log4cxx_appender = 0;
+    g_log4cxx_appender = log4cxx::AppenderPtr();
   }
   // reset this so that the logger doesn't get crashily destroyed
   // again during global destruction.  
   //
   // See https://code.ros.org/trac/ros/ticket/3271
   //
-  log4cxx::Logger::getRootLogger()->getLoggerRepository()->shutdown();
+  static_cast<log4cxx::spi::LoggerRepositoryPtr>(log4cxx::Logger::getRootLogger()->getLoggerRepository())->shutdown();
```

Comment:
this changes seems not to bee needed and produces segfaults with log4cxx 13.

Reply:
Without this change I do get the following error:
```
[ 28%] Building CXX object CMakeFiles/rosconsole_log4cxx.dir/src/rosconsole/impl/rosconsole_log4cxx.cpp.o
/root/ws/src/rosconsole/src/rosconsole/impl/rosconsole_log4cxx.cpp: In function ‘void ros::console::impl::shutdown()’:
/root/ws/src/rosconsole/src/rosconsole/impl/rosconsole_log4cxx.cpp:385:58: error: base operand of ‘->’ has non-pointer type ‘log4cxx::spi::LoggerRepositoryWeakPtr’ {aka ‘std::weak_ptr<log4cxx::spi::LoggerRepository>’}
  385 |   log4cxx::Logger::getRootLogger()->getLoggerRepository()->shutdown();
      |                                                          ^~
make[3]: *** [CMakeFiles/rosconsole_log4cxx.dir/build.make:76: CMakeFiles/rosconsole_log4cxx.dir/src/rosconsole/impl/rosconsole_log4cxx.cpp.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:368: CMakeFiles/rosconsole_log4cxx.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
```

Reply:
Do you have an idea why it is SEGFAULTing?

Reply:
This is with log4cxx 12, I guess?
You can find the version I uploaded for Debian here:
https://salsa.debian.org/science-team/ros-rosconsole/-/blob/dd5d554993aacc6b266e130bd8da5a7b1fdcaaf6/debian/patches/0002-Fix-for-log4cxx-0.12.0.patch
and the one currently used for log4cxx 13 here:
https://salsa.debian.org/science-team/ros-rosconsole/-/tree/master/debian/patches

those got quiet a bit of testing by me and others and there where no problems.

Reply:
> Do you have an idea why it is SEGFAULTing?

Because in log4cxx 0.13 they changed the pointer **to a raw pointer** after switching **to weak_ptr** before in 0.12.
Thus in 0.13 your patch creates a shared_ptr from a raw pointer and deletes the logger repository after evaluating the line without having ownership. I'm not surprised that segfaults.

Please see https://github.com/ros/rosconsole/commit/e3753eec58bf4e76012d019fd307349f94d1d0be for a patch used in ROS-O that's supposed to support log4cxx 0.11 to 0.13. Sadly I had to add another code path to handle 0.12.